### ISSUE TITLE
When is the code expected to be made public?

### ISSUE THREAD

**Issue by xkx10** [hustweiwan/DarkFed#1]
Hello, I am very admired to see your paper, I also do research on federated learning backdoor attacks on the basis of 3dfed's work, but I have never been able to reproduce the backdoor effect in the paper, the problem I often encounter is the scene of loss = NAN after training many times, I hope to refer to your code to get some help. Or can you give some advice?
We look forward to hearing from you, thank you!

**Comment by hustweiwan** [comment]
Thank you for your interest in DarkFed. I did encounter similar difficulties when reproducing 3Fed. For instance, running time became excessively long under DeepSight defense (which is also why DarkFed didn't consider DeepSight). Additionally, I faced the issue of loss=nan while running FLDetector. Here's the explanation **provided by the authors of 3DFed** and I hope it would be helpful for you: 

This problem is a common engineering issue in federated learning backdoor attacks. The cause is that after the backdoor model is aggregated by the aggregator, the running_var parameter of the BatchNormalization layer in the ResNet model becomes negative. However, since running_var needs to be square rooted during data computation, negative values are not allowed, resulting in computation errors showing nan. This issue also exists in the pioneering work on federated backdoor attacks, including the AISTATS paper from 2018. The approach in the Distributed Backdoor Attack (the ICLR paper) is to use a smaller global learning rate when the attacker poisons the data (i.e., reducing the eta parameter in the YAML file), and we also prefer this solution.

We believe another feasible approach is for the attacker to manually slow down the momentum parameter of the BatchNorm2d layer when training their own backdoor model, to prevent the running_var parameter from updating too quickly and causing parameter explosion. If you wish to use this method, you can call the following function before line 57 in the training.py file (i.e., before starting the training of the backdoor model), where the model refers to the attacker's backdoor model:

def adjust_bn_layer(self, model: Module):
    for m in model.modules():
        if isinstance(m, nn.BatchNorm2d):
            m.momentum = 0.01
